### PR TITLE
bugfix

### PR DESCRIPTION
**Pull request by jacobsela**
## What changes are proposed in this pull request?

Make CLIP zoo model work on all GPUs in a system.

## How is this patch tested? If it is not, please explain why.

I ran embeddings on the model on GPUs other than 'coda:0'

## Release Notes
-   [x] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [x] Other - fiftyone.zoo

### CODE REVIEW

**Comment by danielgural** [comment]
LGTM, still works fine and I can see difference between cpu and cuda. Note for future, this change is not pulled upstream by
`fob.compute_similarity(
    dataset,
    model="clip-vit-base32-torch",
    brain_key="img_sim",
    device="cuda",
)
`
and just noticed. Something for next time :)

@harpreetsahota204 can you run this code when you test:
import fiftyone.brain as fob
model = foz.load_zoo_model("clip-vit-base32-torch", device="cuda")
print(model._model.visual.conv1._parameters["weight"][0].device)

To make sure the model is also multi-gpu

**Review by danielgural — DISMISSED** [review]
LGTM 

```py
import fiftyone.brain as fob
from fiftyone import ViewField as F

dataset = foz.load_zoo_dataset('quickstart')
session = fo.launch_app(dataset)
model = foz.load_zoo_model("clip-vit-base32-torch", device="cuda")
embeddings = dataset.compute_embeddings(model)
```

worked as expected

File: fiftyone/utils/clip/zoo.py [hunk 1]
```diff
@@ -187,7 +187,7 @@ def _predict_all(self, imgs):
         frame_size = (width, height)
 
         if self._using_gpu:
-            imgs = imgs.cuda()
+            imgs = imgs.to(torch.device(self.device))
```
Comment (brimoor) [inline fiftyone/utils/clip/zoo.py:190]: Looks like all of the models touched in this PR are subclasses of `TorchImageModel`, for which `self.device` is already a `torch.device()`:

https://github.com/voxel51/fiftyone/blob/84f7e27986a61b9c4b8c9e3dfc1ed032e55df42d/fiftyone/utils/torch.py#L530

So just `.to(self.device)` should suffice I believe?

File: fiftyone/utils/open_clip.py [hunk 2]
```diff
@@ -143,7 +143,7 @@ def _predict_all(self, imgs):
         frame_size = (width, height)
 
         if self._using_gpu:
-            imgs = imgs.cuda()
+            imgs = imgs.to(self.device)
```
[the remark this answers is not in the document]
Comment (brimoor) [reply fiftyone/utils/open_clip.py:146]: @jacobsela coderabbit raises an interesting point here. Does `torch.amp.autocast("cuda")` need to be updated?

[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/open_clip.py:146]: Yeah this is valid and will cause problems if not handled. It's in my todo for this week to more thoroughly review the code before moving further with this PR because this message makes me think that there are probably more places I haven't noticed that make hardware assumptions.

**Comment by brimoor** [comment]
@jacobsela friendly ping here; can we get this completed and merged?

**Comment by deltheil** [comment]
Any chance to modify `fiftyone/utils/transformers.py` in addition (since Hf transformers models are impacted too: see #5271)? Thanks!

**Comment by jacobsela** [comment]
@deltheil That can be done, we'll look into it.

**Comment by jacobsela** [comment]
@brimoor 

Currently testing all zoo models. There seem to be some other unrelated issues that may be worth addressing, e.g. https://github.com/voxel51/fiftyone/issues/5359 and an error pasted below with open clip.

I'll push the fixes for transformers once git goes back up.

This also makes me think that we may need proper testing when adding new zoo models. The code isn't very consistent. Not sure if it's worth the time sink though.

Current status:

Tested devices ['cpu', 'cuda:2']

# Tested models - all pass besides open-clip-torch
-------------
classification-transformer-torch
clip-vit-base32-torch
mnasnet0.5-imagenet-torch
keypoint-rcnn-resnet50-fpn-coco-torch
depth-estimation-transformer-torch
segment-anything-2.1-hiera-tiny-image-torch
resnext101-32x8d-imagenet-torch
yolov5s-coco-torch
mobilenet-v2-imagenet-torch
wide-resnet50-2-imagenet-torch
vgg16-imagenet-torch
densenet121-imagenet-torch
densenet201-imagenet-torch
dinov2-vits14-torch
fcn-resnet50-coco-torch
retinanet-resnet50-fpn-coco-torch
densenet161-imagenet-torch
vgg13-bn-imagenet-torch
segment-anything-2-hiera-large-image-torch
resnet152-imagenet-torch
wide-resnet101-2-imagenet-torch
dinov2-vitb14-torch
vgg16-bn-imagenet-torch
deeplabv3-resnet50-coco-torch
vgg13-imagenet-torch
detection-transformer-torch
fcn-resnet101-coco-torch
squeezenet-imagenet-torch
resnet50-imagenet-torch
squeezenet-1.1-imagenet-torch
yolov5l-coco-torch
vgg11-bn-imagenet-torch
vgg19-bn-imagenet-torch
resnet34-imagenet-torch
shufflenetv2-1.0x-imagenet-torch
faster-rcnn-resnet50-fpn-coco-torch
resnet18-imagenet-torch
resnext50-32x4d-imagenet-torch
mnasnet1.0-imagenet-torch
alexnet-imagenet-torch
yolov5x-coco-torch
vgg11-imagenet-torch
mask-rcnn-resnet50-fpn-coco-torch
segment-anything-2.1-hiera-large-image-torch
segment-anything-2-hiera-small-image-torch
googlenet-imagenet-torch
densenet169-imagenet-torch
inception-v3-imagenet-torch
segment-anything-2.1-hiera-small-image-torch
segmentation-transformer-torch
dinov2-vitg14-torch
resnet101-imagenet-torch
segment-anything-2-hiera-base-plus-image-torch
shufflenetv2-0.5x-imagenet-torch
segment-anything-2.1-hiera-base-plus-image-torch
yolov5m-coco-torch
deeplabv3-resnet101-coco-torch
dinov2-vitl14-torch
yolov5n-coco-torch
open-clip-torch
vgg19-imagenet-torch
segment-anything-2-hiera-tiny-image-torch
=============

# Not tested models (I need to setup an environment to test all of these):

Key                                                          Value                                                       
------------------------------------------------------------ ------------------------------------------------------------
med-sam-2-video-torch                                        Model is not an image model                                 
rtdetr-l-coco-torch                                          Model does not have a device attribute                      
rtdetr-x-coco-torch                                          Model does not have a device attribute                      
segment-anything-2-hiera-base-plus-video-torch               Model is not an image model                                 
segment-anything-2-hiera-large-video-torch                   Model is not an image model                                 
segment-anything-2-hiera-small-video-torch                   Model is not an image model                                 
segment-anything-2-hiera-tiny-video-torch                    Model is not an image model                                 
segment-anything-2.1-hiera-base-plus-video-torch             Model is not an image model                                 
segment-anything-2.1-hiera-large-video-torch                 Model is not an image model                                 
segment-anything-2.1-hiera-small-video-torch                 Model is not an image model                                 
segment-anything-2.1-hiera-tiny-video-torch                  Model is not an image model                                 
segment-anything-vitb-torch                                  Failed to load model                                        
segment-anything-vith-torch                                  Failed to load model                                        
segment-anything-vitl-torch                                  Failed to load model                                        
yolo-nas-torch                                               Failed to load model                                        
yolo11l-coco-torch                                           Model does not have a device attribute                      
yolo11l-seg-coco-torch                                       Model does not have a device attribute                      
yolo11m-coco-torch                                           Model does not have a device attribute                      
yolo11m-seg-coco-torch                                       Model does not have a device attribute                      
yolo11n-coco-torch                                           Model does not have a device attribute                      
yolo11n-seg-coco-torch                                       Model does not have a device attribute                      
yolo11s-coco-torch                                           Model does not have a device attribute                      
yolo11s-seg-coco-torch                                       Model does not have a device attribute                      
yolo11x-coco-torch                                           Model does not have a device attribute                      
yolo11x-seg-coco-torch                                       Model does not have a device attribute                      
yolov10l-coco-torch                                          Model does not have a device attribute                      
yolov10m-coco-torch                                          Model does not have a device attribute                      
yolov10n-coco-torch                                          Model does not have a device attribute                      
yolov10s-coco-torch                                          Model does not have a device attribute                      
yolov10x-coco-torch                                          Model does not have a device attribute                      
yolov8l-coco-torch                                           Model does not have a device attribute                      
yolov8l-obb-dotav1-torch                                     Model does not have a device attribute                      
yolov8l-oiv7-torch                                           Model does not have a device attribute                      
yolov8l-seg-coco-torch                                       Model does not have a device attribute                      
yolov8l-world-torch                                          Model does not have a device attribute                      
yolov8m-coco-torch                                           Model does not have a device attribute                      
yolov8m-obb-dotav1-torch                                     Model does not have a device attribute                      
yolov8m-oiv7-torch                                           Model does not have a device attribute                      
yolov8m-seg-coco-torch                                       Model does not have a device attribute                      
yolov8m-world-torch                                          Model does not have a device attribute                      
yolov8n-coco-torch                                           Model does not have a device attribute                      
yolov8n-obb-dotav1-torch                                     Model does not have a device attribute                      
yolov8n-oiv7-torch                                           Model does not have a device attribute                      
yolov8n-seg-coco-torch                                       Model does not have a device attribute                      
yolov8s-coco-torch                                           Model does not have a device attribute                      
yolov8s-obb-dotav1-torch                                     Model does not have a device attribute                      
yolov8s-oiv7-torch                                           Model does not have a device attribute                      
yolov8s-seg-coco-torch                                       Model does not have a device attribute                      
yolov8s-world-torch                                          Model does not have a device attribute                      
yolov8x-coco-torch                                           Model does not have a device attribute                      
yolov8x-obb-dotav1-torch                                     Model does not have a device attribute                      
yolov8x-oiv7-torch                                           Model does not have a device attribute                      
yolov8x-seg-coco-torch                                       Model does not have a device attribute                      
yolov8x-world-torch                                          Model does not have a device attribute                      
yolov9c-coco-torch                                           Model does not have a device attribute                      
yolov9c-seg-coco-torch                                       Model does not have a device attribute                      
yolov9e-coco-torch                                           Model does not have a device attribute                      
yolov9e-seg-coco-torch                                       Model does not have a device attribute                      
========================================================================================================================

# Errors

======================================================================
ERROR: test_all_torch_image_models (__main__.TestDeviceUsage) (model_name='open-clip-torch', device='cpu', input_format='numpy')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jacobs/test_device_usage.py", line 86, in _test_image_model
    fo_torch_model.predict_all(dummy_inputs)
  File "/home/jacobs/fiftyone/fiftyone/utils/torch.py", line 691, in predict_all
    return self._predict_all(imgs)
  File "/home/jacobs/fiftyone/fiftyone/utils/open_clip.py", line 137, in _predict_all
    imgs = [self._preprocess(img).unsqueeze(0) for img in imgs]
  File "/home/jacobs/fiftyone/fiftyone/utils/open_clip.py", line 137, in <listcomp>
    imgs = [self._preprocess(img).unsqueeze(0) for img in imgs]
TypeError: 'bool' object is not callable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jacobs/test_device_usage.py", line 147, in test_all_torch_image_models
    self._test_image_model(fo_torch_model, model_name, device, input_format=input_format)
  File "/home/jacobs/test_device_usage.py", line 90, in _test_image_model
    raise Exception(f"Failed to run model {model_name} on device {device}") from e
Exception: Failed to run model open-clip-torch on device cpu

======================================================================
ERROR: test_all_torch_image_models (__main__.TestDeviceUsage) (model_name='open-clip-torch', device='cuda:2', input_format='numpy')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jacobs/test_device_usage.py", line 86, in _test_image_model
    fo_torch_model.predict_all(dummy_inputs)
  File "/home/jacobs/fiftyone/fiftyone/utils/torch.py", line 691, in predict_all
    return self._predict_all(imgs)
  File "/home/jacobs/fiftyone/fiftyone/utils/open_clip.py", line 137, in _predict_all
    imgs = [self._preprocess(img).unsqueeze(0) for img in imgs]
  File "/home/jacobs/fiftyone/fiftyone/utils/open_clip.py", line 137, in <listcomp>
    imgs = [self._preprocess(img).unsqueeze(0) for img in imgs]
TypeError: 'bool' object is not callable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jacobs/test_device_usage.py", line 147, in test_all_torch_image_models
    self._test_image_model(fo_torch_model, model_name, device, input_format=input_format)
  File "/home/jacobs/test_device_usage.py", line 90, in _test_image_model
    raise Exception(f"Failed to run model {model_name} on device {device}") from e
Exception: Failed to run model open-clip-torch on device cuda:2

----------------------------------------------------------------------
Ran 1 test in 275.831s

FAILED (errors=2)

**Comment by jacobsela** [comment]
# Still no testing done for MPS

# Models tested w/ various coda devices
alexnet-imagenet-torch
classification-transformer-torch
clip-vit-base32-torch
deeplabv3-resnet101-coco-torch
deeplabv3-resnet50-coco-torch
densenet121-imagenet-torch
densenet161-imagenet-torch
densenet169-imagenet-torch
densenet201-imagenet-torch
depth-estimation-transformer-torch
detection-transformer-torch
dinov2-vitb14-torch
dinov2-vitg14-torch
dinov2-vitl14-torch
dinov2-vits14-torch
faster-rcnn-resnet50-fpn-coco-torch
fcn-resnet101-coco-torch
fcn-resnet50-coco-torch
googlenet-imagenet-torch
inception-v3-imagenet-torch
keypoint-rcnn-resnet50-fpn-coco-torch
mask-rcnn-resnet50-fpn-coco-torch
mnasnet0.5-imagenet-torch
mnasnet1.0-imagenet-torch
mobilenet-v2-imagenet-torch
open-clip-torch
resnet101-imagenet-torch
resnet152-imagenet-torch
resnet18-imagenet-torch
resnet34-imagenet-torch
resnet50-imagenet-torch
resnext101-32x8d-imagenet-torch
resnext50-32x4d-imagenet-torch
retinanet-resnet50-fpn-coco-torch
rtdetr-l-coco-torch
rtdetr-x-coco-torch
segment-anything-2-hiera-base-plus-image-torch
segment-anything-2-hiera-large-image-torch
segment-anything-2-hiera-small-image-torch
segment-anything-2-hiera-tiny-image-torch
segment-anything-2.1-hiera-base-plus-image-torch
segment-anything-2.1-hiera-large-image-torch
segment-anything-2.1-hiera-small-image-torch
segment-anything-2.1-hiera-tiny-image-torch
segmentation-transformer-torch
shufflenetv2-0.5x-imagenet-torch
shufflenetv2-1.0x-imagenet-torch
squeezenet-1.1-imagenet-torch
squeezenet-imagenet-torch
vgg11-bn-imagenet-torch
vgg11-imagenet-torch
vgg13-bn-imagenet-torch
vgg13-imagenet-torch
vgg16-bn-imagenet-torch
vgg16-imagenet-torch
vgg19-bn-imagenet-torch
vgg19-imagenet-torch
wide-resnet101-2-imagenet-torch
wide-resnet50-2-imagenet-torch
yolo11l-coco-torch
yolo11l-seg-coco-torch
yolo11m-coco-torch
yolo11m-seg-coco-torch
yolo11n-coco-torch
yolo11n-seg-coco-torch
yolo11s-coco-torch
yolo11s-seg-coco-torch
yolo11x-coco-torch
yolo11x-seg-coco-torch
yolov10l-coco-torch
yolov10m-coco-torch
yolov10n-coco-torch
yolov10s-coco-torch
yolov10x-coco-torch
yolov5l-coco-torch
yolov5m-coco-torch
yolov5n-coco-torch
yolov5s-coco-torch
yolov5x-coco-torch
yolov8l-coco-torch
yolov8l-obb-dotav1-torch
yolov8l-oiv7-torch
yolov8l-seg-coco-torch
yolov8l-world-torch
yolov8m-coco-torch
yolov8m-obb-dotav1-torch
yolov8m-oiv7-torch
yolov8m-seg-coco-torch
yolov8m-world-torch
yolov8n-coco-torch
yolov8n-obb-dotav1-torch
yolov8n-oiv7-torch
yolov8n-seg-coco-torch
yolov8s-coco-torch
yolov8s-obb-dotav1-torch
yolov8s-oiv7-torch
yolov8s-seg-coco-torch
yolov8s-world-torch
yolov8x-coco-torch
yolov8x-obb-dotav1-torch
yolov8x-oiv7-torch
yolov8x-seg-coco-torch
yolov8x-world-torch
yolov9c-coco-torch
yolov9c-seg-coco-torch
yolov9e-coco-torch
yolov9e-seg-coco-torch
zero-shot-classification-transformer-torch
zero-shot-detection-transformer-torch

# Models that are still problematic
yolov5 - loads on the coda:0 before being loaded to the device in the argument. Not sure why
hugging face zero shot transformers - expect `input_ids` argument that isn't passed. probably unrelated error.
open-clip-torch - self._preprocess is a bool instead of a callable. not sure why. probably unrelated.

# models that haven't been tested - need to setup env
med-sam-2-video-torch -  Model is not an image model                                 
segment-anything-2-hiera-base-plus-video-torch - Model is not an image model                                 
segment-anything-2-hiera-large-video-torch - Model is not an image model                                 
segment-anything-2-hiera-small-video-torch - Model is not an image model                                 
segment-anything-2-hiera-tiny-video-torch - Model is not an image model                                 
segment-anything-2.1-hiera-base-plus-video-torch - Model is not an image model                                 
segment-anything-2.1-hiera-large-video-torch - Model is not an image model                                 
segment-anything-2.1-hiera-small-video-torch - Model is not an image model                                 
segment-anything-2.1-hiera-tiny-video-torch - Model is not an image model                                 
segment-anything-vitb-torch - Failed to load model                                        
segment-anything-vith-torch - Failed to load model                                        
segment-anything-vitl-torch - Failed to load model                                        
yolo-nas-torch - Failed to load model

File: fiftyone/utils/ultralytics.py [hunk 3]
```diff
@@ -20,6 +20,7 @@
 import fiftyone.zoo.models as fozm
 
 ultralytics = fou.lazy_import("ultralytics")
+torch = fou.lazy_import("torch")
```
Comment (kaixi-wang) [inline fiftyone/utils/ultralytics.py:23]: Torch isn't a requirement. Will this cause problems or does any code that uses ultralytics require torch already.

Comment (jacobsela) [reply fiftyone/utils/ultralytics.py:23]: Yes, UItralytics depends on Torch.

**Comment by brimoor** [comment]
Adding @manushreegangwar and @mwoodson1 as ML team reviewers 😄

**Comment by brimoor** [comment]
@jacobsela can you rebase on latest `develop`? Looks like these is a merge conflict that would currently prevent merging this.

Also:

> yolov5 - loads on the coda:0 before being loaded to the device in the argument. Not sure why

We're [using Ultralytics' model](https://github.com/voxel51/fiftyone/blob/c5d3bf5715f1f5fa1049ff381f2a3508dd31df26/fiftyone/zoo/models/manifest-torch.json#L2964-L2976) here. Can anything be done to address this?

> hugging face zero shot transformers - expect input_ids argument that isn't passed. probably unrelated error.

On `develop` on macOS with CPU, these work for me. Are you seeing something different?
- https://docs.voxel51.com/model_zoo/models.html#detection-transformer-torch
- https://docs.voxel51.com/model_zoo/models.html#classification-transformer-torch

> open-clip-torch - self._preprocess is a bool instead of a callable. not sure why. probably unrelated.

On `develop` on macOS with CPU, this works for me. Are you seeing something different?
- https://docs.voxel51.com/model_zoo/models.html#open-clip-torch

**Comment by jacobsela** [comment]
Resolved the yolo5 issue. When loading from torch hub, it will automatically load the model onto the currently set default device. Wrapping line 788 in `fiftyone.utils.torch` in a `with torch.device("cpu")` fixes this while maintaining the default device in other parts of the code.

**Comment by jacobsela** [comment]
I'm just going to pass "cpu" to always be the device in the manifest. model is moved to correct device afterwards.

**Comment by jacobsela** [comment]
status:

1. Rebased locally and force pushed, `git merge --no-commit --no-ff bugfix/zoo-clip-support-for-multi-gpu-setups` from an up-to-date `develop` gives no errors. I'm not sure if this was the proper way of doing this.
2. Fixed the yolov5 issue by updating manifests to by default use CPU. In general it seems that there is no way to pass arguments directly into the `entrypoint_fcn` of a zoo model from the code, just the manifest. Am I missing something?
3. For the zero-shot transformers I'm getting the same error `ValueError: You have to specify input_ids` on cpu in develop.  Something may be broken on my env.
4. SAM2 models have some internal component that defaults to loading on cuda:0. I don't know if it's just my environment or a general thing. I don't know if this is worth the trouble of debugging. Assuming the user has a functioning default "cuda" device with enough free memory they probably wouldn't even notice.

TL;DR
Models that are still up in the air:
- video models
- yolo-nas (getting error when downloading weights)
- open-clip-torch (can't get it to run)
- zero shot transformers (can't get them to run)
Works but for whatever reason loads on "cuda" before going to desired device:
- sam 2 models

**Comment by danielgural** [comment]
MPS works on all but some transformers due to an aten::upsample_bicubic2d.out operator. Error spits out correctly as "not supported on MPS yet" from torch. 

Multi GPU works except for zero-shot-classification-transformer-torch on device cuda

```
Traceback (most recent call last):
  File "/home/dan/model_testing/jacob_test.py", line 158, in test_all_torch_image_models
    self._test_image_model(fo_torch_model, model_name, device, input_format=input_format)
  File "/home/dan/model_testing/jacob_test.py", line 90, in _test_image_model
    raise Exception(f"Failed to run model {model_name} on device {device}") from e
Exception: Failed to run model zero-shot-classification-transformer-torch on device cuda
```

+1 to clip input_ids issue. 

LGTM for my tests just needs the stated above fixes

**Comment by jacobsela** [comment]
# open-clip update

The preprocessor of open clip is loaded in line 92 of `fiftyone.utils.open_clip` in the method `_load_model` of the class `TorchOpenClipModel`.

This method is called by `fiftyone.utils.torch.TorchImageModel` in the `__init__` in line 538. It is then set to `True` a few lines later in line 544.

Even when this issue is fixed by saving the preprocessor in an auxiliary variable and setting `self.preprocess` at the end of `TorchOpenClipModel`'s `__init__`, `self.predict` doesn't work with any of the possible input types in the contract defined by `TorchImageModel` (np array, pil image, torch tensor). The actual open clip model expects PIL image input ruling out numpy and torch inputs.

Even when `self.predict` is given PIL image input, the `unsqueeze(0)` in line 138 of `fiftyone.utils.open_clip` followed by the `stack` in line 141 creates an extra dimension that causes an error.

Fixing all of these issues fixes the problem in my env.

**Comment by jacobsela** [comment]
@danielgural @mwoodson1 @manushreegangwar Given the fact that most models are working, I suggest we currently merge this PR as it is and open separate tickets for the other issues, e.g. https://github.com/voxel51/fiftyone/pull/5395

Let me know what you think.

**Comment by jacobsela** [comment]
Leaving this test script here in case we need it again: 
https://github.com/voxel51/test_zoo_models_script

**Comment by brimoor** [comment]
@jacobsela can you retarget this PR at `release/v1.3.0` so we can include these fixes in the next release?

**Review by manushreegangwar — APPROVED** [review]
LGTM

**Comment by rromanchuk** [comment]
Something just regressed after i updated. I haven't investigated this PR yet, probably user error but the fact i would assume these changes to be a NOOP, because i'm not using the model zoo, makes me more suspicious. 

```python
model = YOLO('best.pt')
model.to("mps")
dataset.apply_model(model, batch_size=10)
```

**Comment by rromanchuk** [comment]
Actually yeah, i was able to test this by loading through `fiftyone.zoo` via `custom_model.json` and now she's flying.

```python
model = foz.load_zoo_model("load-with-zoo", device="mps", install_requirements=True)
dataset.apply_model(model, batch_size=10, device="mps")
```


```json
{
  "models": [
      {
          "base_name": "load-with-zoo",
          "base_filename": "best.pt",
          "version": null,
          "description": "Test loading through load_zoo_model",
          "manager": {
                "type": "fiftyone.core.models.ModelManager",
                "config": {
                    "url": "file:///Volumes/foo/bar/best.pt"
                }
            },
          "default_deployment_config_dict": {
              "type": "fiftyone.utils.ultralytics.FiftyOneYOLODetectionModel",
              "config": {
                "device": "mps"
              }
          },
          "requirements": {
             "packages": [
                    "torch>=1.7.0",
                    "torchvision>=0.8.1",
                    "ultralytics>=8.2.0"
                ],
                "cpu": {
                    "support": true
                },
                "gpu": {
                    "support": true
                }
          }
      }
  ]
}
```
